runtime/include/cloe/data_broker.hpp

```diff
@@ -0,0 +1,1055 @@
+                            signal->name())
+              : fmt::format("The name for a signal must not be an empty string"));
+    }
+    // Mutate signals
+    std::pair<signal_container_t::iterator, bool> inserted =
+        signals_.try_emplace(std::string(new_name), std::move(signal));
+    if (!inserted.second) {
+      throw std::invalid_argument(signal->names().size()
+                                      ? fmt::format("Unable to alias signal '{}' to '{}' since a "
+                                                    "signal with the name already exists.",
+                                                    signal->name(), new_name)
+                                      : fmt::format("Unable to create a new signal '{}' since a "
+                                                    "signal with the name already exists.",
+                                                    new_name));
+    }
+    // signals mutated, there is a liability in case of exceptions
+    try {
+      signal_ptr_t result = inserted.first->second;
+      result->add_name(new_name);
+      return result;
+    } catch (...) {
+      // fullfill exception guarantee (map.erase(iter) does not throw)
+      signals_.erase(inserted.first);
+      throw;
+    }
+  }
+  /**
+    * Gives an existing signal a (new) name
+    * \param old_name Name of the existing signal
+    * \param new_name New name of the signal
+    * \param f flag_type flags used to guide the interpretation of the character sequence as a regular expression
+    * \return Pointer to the signal
+    * \note If an exception is thrown by any operation, the aliasing has no effect.
+    */
+  signal_ptr_t alias(std::string_view old_name, std::string_view new_name,
+                     std::regex::flag_type f = std::regex_constants::ECMAScript) {
+    std::regex regex = std::regex(std::string(old_name), f);
+    auto it1 = signals().begin();
+    auto it2 = signals().begin();
+    auto end = signals().end();
+    const auto predicate = [&](const auto& item) -> bool {
+      std::smatch match;
+      return std::regex_match(item.first, match, regex);
+    };
+    it1 = (it1 != end) ? std::find_if(it1, end, predicate) : end;
+    it2 = (it1 != end) ? std::find_if(std::next(it1), end, predicate) : end;
+    if (it2 != end) {
+      throw std::invalid_argument(fmt::format(
+          "Regex pattern matches more than one signal. First match: '{}' Second Match '{}'",
+          it1->first, it2->first));
+    }
+    if (it1 == end) {
+      throw std::invalid_argument(
+          fmt::format("Regex pattern matches zero signals. Pattern: '{}'", old_name));
+      return nullptr;
+    }
+    signal_ptr_t result = alias(it1->second, new_name);
+    return result;
+  }
+
+  /**
+    * Declares a new signal
+    * \tparam T Type of the signal
+    * \param name Name of the signal
+    * \return Pointer to the specified signal
+    */
+  template <typename T>
+  signal_ptr_t declare(const std::string_view new_name) {
```

Comment:
If performance is important, we should use `string` and not `string_view` for methods that will store a new string. This potentially saves us an allocation on the heap. `std::string_view` is still better than `const std::string` though.

Here and in many other methods.

Reply:
We changed this intentionally to remove the overloads.

Reply:
My point is: syscalls and allocations are expensive. string_view is explicitely for use-cases where you only want a read-only view into strings. Here we are are going to put string into the Signal, so we can take a string and move it. That's cheaper. You can still call it with a string_view if you like, but you can also move a string into it.

Reply:
There's a great video on more or less this topic which is motivation for my comment: https://www.youtube.com/watch?v=PNRju6_yn3o

TLDR: It's more or less on what the best way is to deal with passing strings to a constructor. Since the string we're talking about will end up in the constructor of a string, I deemed it as relevant.

Reply:
Here's what I took from the talk:
```cpp
#include <string>
#include <string_view>
using namespace std::literals;

struct Foo {
    std::string name;

    void set_name(std::string s) { name = std::move(s); }
};

struct Bar {
    std::string_view name;

    void set_name(std::string_view s) { name = s; }
};

int main() {

    Foo f;
    f.set_name("const"); // const char* implicitely converted to string: 1 construct, 1 move
    f.set_name(std::string("string view"sv)); // string_view converted to string: 1 construct, 1 move
    f.set_name("string"s); // rvalue moved into argument: 1 construct, 1 move
    std::string str1 = "owning";
    f.set_name(std::move(str1)); // existing string moved into argument: 1 move
    // Foo: 4 constructions, 4 moves

    Bar b;
    b.set_name("const"); // const char* implicitely converted to string_view: 1 construct
    b.set_name("string view"sv); // string_view passed: 1 construct
    b.set_name("string"s); // rvalue string converted to string_view: 2 construct
    std::string str2 = "owning";
    b.set_name(std::move(str2)); // existing string moved into argument: 2 construct
    // Bar: 6 constructions, 0 moves
}
```

Allocations are orders of magnitude more expensive than moves, so despite always involving a move, Josuttis recommended to use `std::string` for most use-cases. The only way to get the most efficient solution each time is to provide overloads (or a template function) for each type.

Reply:
So funny story: I watched the talk from Herb Sutter from CppCon14 last night, and my code example above is very similar to his example but he recommends `const std::string&` — because he benchmarks calling the same function ~50 times over. Turns out a copy only allocates if the string doesn't already have enough memory allocated.

Thus my argumentation here only really holds for setting empty strings or when constructing new ones.

I think in `declare` it's still best to use `std::string`, but just wanted to point out in the end if we care about performance, we need to measure. Best to aim for correctness and simplicity first, and fine-tune performance if necessary.

Reply:
Crazy :D strings are funny beasts with short string optimization and the likes. Fully agree though, benchmarking is key in the end.